File: manma/box_storage.py
from .box import Box
from .rbt import RedBlackTree, NIL
from .heap import MaxHeap, MinHeap


class SideNode(object):
    def __init__(self, side):
        self.side = side
        self.heights = RedBlackTree()

    def __lt__(self, other):  # This is like C#s CompareTo
        return self.side < other

    def __eq__(self, other):
        return self.side == other

    def __gt__(self, other):
        return self.side > other


class BoxStorage(object):
    def __init__(self):  # O(1)
        self.sides = RedBlackTree()
        self.boxes_max_heap = MaxHeap()
        self.boxes_min_heap = MinHeap()

    def insert_box(self, side, height):  # O(lgn) + O(lgm) -> O(lgnm)
        # Red Black Trees  # O(lgn) + O(lgm) -> O(lg(nm))
        side_node = self.sides.search(side)

        if side_node == NIL:  # The side we want does not exist
            side_node = SideNode(side)
            side_node.heights.insert(height)
            self.sides.insert(side_node)

        else:
            side_node.key.heights.insert(height)

        # Heaps O(lg(max(n, m)))
        box = Box(side, height)

        if self.boxes_max_heap.size == 0 or box < self.boxes_max_heap.top():
            self.boxes_max_heap.insert(box)

        else:
            self.boxes_min_heap.insert(box)

        if self.boxes_max_heap.size - self.boxes_min_heap.size > 1:
            self.boxes_min_heap.insert(self.boxes_max_heap.extract())

        if self.boxes_min_heap.size - self.boxes_max_heap.size > 1:
            self.boxes_max_heap.insert(self.boxes_min_heap.extract())

    def remove_box(self, side, height):  # O(max(n, m))
        # Red Black Trees: O(lgm) + O(lgn) -> O(lgnm)
        side_node = self.sides.search(side)
        if side_node == NIL:
            raise KeyError("Can't remove box, box does not exist")

        height_node = side_node.key.heights.search(height)
        if height_node == NIL:
            raise KeyError("Can't remove box, box does not exist")

        side_node.key.heights.delete(height_node)

        if side_node.key.heights.root == NIL:
            self.sides.delete(side_node)

        # Heaps: Couldn't find a way to find it that's better than O(max(n, m)
        volume = Box(side, height).volume()

        for index in range(1, len(self.boxes_min_heap.array) + 1):
            item = self.boxes_min_heap.item_at(index)
            if item == volume:
                self.boxes_min_heap.delete_item_at_index(index)
                return

        for index in range(1, len(self.boxes_max_heap.array) + 1):
            item = self.boxes_max_heap.item_at(index)
            if item == volume:
                self.boxes_max_heap.delete_item_at_index(index)
                return

    def get_box(self, side, height) -> Box:  # O(mlgn)
        """Return the box with minimal volume that answers given criteria."""
        all_side_nodes = [side_node
                          for side_node
                          in self.sides.inorder_walk()
                          if side_node.key >= side]

        minimal_volume = float("infinity")
        minimal_volume_side = 0
        minimal_volume_height = 0

        for side_node in all_side_nodes:
            minimal_height = side_node.heights.search_successor(height)
            volume = Box(side_node.key, minimal_height).volume()

            if volume < minimal_volume:
                minimal_volume = volume
                minimal_volume_height = minimal_height
                minimal_volume_side = side_node.key

        if minimal_volume == float("inifinity"):
            raise KeyError("No such box exists")

        return Box(minimal_volume_side, minimal_volume_height)

    def check_box(self, side, height) -> bool:  # O(mlgn)
        all_side_nodes = [side_node
                          for side_node
                          in self.sides.inorder_walk()
                          if side_node.key >= side]

        if len(all_side_nodes) == 0:
            return False

        for side_node in all_side_nodes:
            if side_node.search_successor(height) != "nil":
                return True

        return False

    def get_median_box(self) -> Box:  # O(1)
        """Return the box with the median volume."""
        return self.boxes_max_heap.top()

File: Maman15/src/BoxStorage.java
public class BoxStorage {

    private RedBlackTree<SideNode> sides;

    /**
     * The BoxStorage class represent the storage of the workshop.
     * We choose to use RedBlackTree to keep all the available sides.
     * Each node contain a RedBlackTree of the available heights.
     */
    BoxStorage(){
        this.sides = new RedBlackTree<>();
    }

    // @param: side and height
    // insert the box to the storage. If the side already exists we add the new height to his heights tree.
    //O(lgnm)
    public void insertBox(double side, double height){
        System.out.println("Insert new box - height: " + height + " ,side: " + side);

        SideNode sideNode = new SideNode(side);
        RedBlackNode<SideNode> existedSideNode = this.sides.search(sideNode);

        if(existedSideNode == null) // The side we want does not exist
        {
            sideNode.heights.insert(height);
            this.sides.insert(sideNode);
        }
        else {
            existedSideNode.key.heights.insert(height);
        }
    }

    // @param: side and height
    // remove the box from the storage.
    //O(lgnm)
    public void removeBox(double side, double height){
        System.out.println("Remove box - height: " + height + " ,side: " + side);

        SideNode sideNode = new SideNode(side);
        RedBlackNode<SideNode> existedSideNode = this.sides.search(sideNode);
        existedSideNode.key.heights.remove(existedSideNode.key.heights.search(height));
    }

    // @param: side and height
    // @return: the mach box
    //O(mlgn)
    public Box getBox(double side, double height){
        System.out.println("Get box - height: " + height + " ,side: " + side);

        SideNode sideNode = new SideNode(side);

        double minimalVolume = Double.MAX_VALUE;
        double minimalVolumeSide = 0;
        double minimalVolumeHeight = 0;

        for (SideNode matchSideNode : this.sides.getGreaterOrEqualsThan(sideNode, this.sides.size())){

            RedBlackNode<Double> minimalHeightNode = matchSideNode.heights.serachSuccessor(height);
            if(minimalHeightNode == null){
                continue;
            }

            double volume = matchSideNode.side * matchSideNode.side * minimalHeightNode.key;
            if(volume < minimalVolume){
                minimalVolume = volume;
                minimalVolumeHeight = minimalHeightNode.key;
                minimalVolumeSide = matchSideNode.side;
            }
        }

        System.out.println("Returned box - height: " + minimalVolumeHeight + " ,side: " + minimalVolumeSide);

        return new Box(minimalVolumeSide,minimalVolumeHeight);
    }

    // @param: side and height
    // @return: true if mach box exists, otherwise false
    //O(mlgn)
    public boolean checkBox(double side, double height) {
        System.out.println("Check if box exists - height: " + height + " ,side: " + side);

        SideNode sideNode = new SideNode(side);
        for (SideNode matchSideNode : this.sides.getGreaterOrEqualsThan(sideNode, this.sides.size())) {
            if (matchSideNode.heights.serachSuccessor(height) != null )
            {
                System.out.println("The box exists");
                return true;
            }
        }

        System.out.println("The box doesn't exist");
        return false;
    }
}
